### PR TITLE
feat: unify to single entry point for keeper exec

### LINKED ISSUE #12: feat: single method entry for gelato
to trigger in one go either:

1. normal topup
2. bpt topup

### PR DESCRIPTION
**Pull request by petrovska-petro**
close #12

### CODE REVIEW

File: test/TopupBptTest.t.sol [hunk 1]
```diff
@@ -23,7 +23,7 @@ contract TopupBptTest is BaseFixture {
 
         // since initially it was minted 1000 it should be way above the buffer
         assertTrue(canExec);
-        assertEq(selector, EXEC_TOP_UP_SELECTOR);
+        assertEq(selector, ADJUST_POOL_SELECTOR);
```
[the remark this answers is not in the document]
Comment (gosuto-inzasheru) [reply test/TopupBptTest.t.sol:26]: i guess we would need to decode `execPayload` and retrieve `PoolAction` -_-

File: test/TopupEUReTest.t.sol [hunk 2]
```diff
@@ -43,26 +45,28 @@ contract TopupTest is BaseFixture {
         delayModule.executeNextTx(EURE, 0, payload, Enum.Operation.Call);
 
         (bool canExec, bytes memory execPayload) = roboModule.checker();
+        (bytes memory dataWithoutSelector, bytes4 selector) = _extractEncodeDataWithoutSelector(execPayload);
 
         assertTrue(canExec);
```
Comment (gosuto-inzasheru) [inline test/TopupEUReTest.t.sol:50]: shouldnt we also check for `assertFalse(canExec)` at the beginning of the test, to make the assert on L50 meaningful?

File: test/TopupEUReTest.t.sol [hunk 3]
```diff
@@ -43,26 +45,28 @@ contract TopupTest is BaseFixture {
...
-        assertEq(bytes4(execPayload), SAFE_TOP_UP_SELECTOR);
+        assertEq(selector, ADJUST_POOL_SELECTOR);
 
         uint256 initialEureBal = IERC20(EURE).balanceOf(GNOSIS_SAFE);
 
         vm.prank(TOP_UP_AGENT);
-        (bool success, bytes memory data) = address(roboModule).call(execPayload);
-        assertTrue(success);
+        (RoboSaverVirtualModule.PoolAction _action, address _card, uint256 _amount) =
+            abi.decode(dataWithoutSelector, (RoboSaverVirtualModule.PoolAction, address, uint256));
```
Comment (gosuto-inzasheru) [inline test/TopupEUReTest.t.sol:57]: aah you decode here! so lets assert `_action` is a pool withdrawal here?

File: test/TopupEUReTest.t.sol [hunk 4]
```diff
@@ -73,6 +73,6 @@ contract TopupTest is BaseFixture {
 
         tokenAmountTargetToMove_ = eureBalance - maxRefill + 100e18;
 
-        roboModule.transferErc20(EURE, tokenAmountTargetToMove_, WETH);
+        // roboModule.transferErc20(EURE, tokenAmountTargetToMove_, WETH);
```
[the remark this answers is not in the document]
Comment (gosuto-inzasheru) [reply test/TopupEUReTest.t.sol]: nvm this, this was already a helper and not a test.

i will work on rewording to make the naming more intuitive to its current functionality

[the remark this answers is not in the document]
Comment (petrovska-petro) [reply test/TopupEUReTest.t.sol]: 5a8fd14760165c4147684b67d11c6d6ae0a3f439

File: test/TopupBptTest.t.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (petrovska-petro) [reply test/TopupBptTest.t.sol]: 5f1fcea904ca397fb0a2d7390dd29c269adb5241

File: test/TopupEUReTest.t.sol [hunk 5]
```diff
@@ -27,12 +29,12 @@ contract TopupTest is BaseFixture {
         (canExec, execPayload) = roboModule.checker();
 
         assertTrue(canExec);
-        assertEq(bytes4(execPayload), SAFE_TOP_UP_SELECTOR);
+        assertEq(bytes4(execPayload), ADJUST_POOL_SELECTOR);
```
[the remark this answers is not in the document]
Comment (petrovska-petro) [reply test/TopupEUReTest.t.sol]: 5f1fcea904ca397fb0a2d7390dd29c269adb5241

File: test/TopupEUReTest.t.sol (hunk 3 above)
Comment (petrovska-petro) [reply test/TopupEUReTest.t.sol]: done in 5f1fcea904ca397fb0a2d7390dd29c269adb5241

File: test/TopupEUReTest.t.sol (hunk 2 above)
Comment (gosuto-inzasheru) [reply test/TopupEUReTest.t.sol]: what i mean here is that L50 in `testExitPool` checks for `canExec` being true, but it doesnt check the state of that var at the start of the function. in other words, maybe `canExec` was true all along. 

if the following were to be included in the beginning of the test, we would actually _prove_ that the transfer out was the thing that was responsible for the state change of `canExec`:
```
(bool initialCanExec,) = roboModule.checker();
assertFalse(initialCanExec);
```

File: src/RoboSaverVirtualModule.sol [hunk 6]
```diff
@@ -35,147 +47,146 @@ contract RoboSaverVirtualModule {
...
-        uint256 txNonce = delayModule.txNonce();
-
-        if (txNonce != queueNonce) {
-            uint256 txQueuedAt = delayModule.getTxCreatedAt(queueNonce - 1);
-            // @note triggers the condition for exec the pendant tx in the delay module
-            if (block.timestamp - txQueuedAt >= delayModule.txCooldown()) {
-                return (true, abi.encodeWithSelector(IDelayModifier.executeNextTx.selector));
-            }
-
-            return (false, bytes("Tx cooldown not reached"));
```
Comment (gosuto-inzasheru) [inline src/RoboSaverVirtualModule.sol:91]: this logic should remain afaik!!!

**Review by gosuto-inzasheru — CHANGES_REQUESTED** [review]

**Comment by gosuto-inzasheru** [comment]
looks like forked block height is again too far away? can we maybe fork on latest by default for now?

File: src/RoboSaverVirtualModule.sol (hunk 6 above)
Comment (petrovska-petro) [reply src/RoboSaverVirtualModule.sol:91]: this logic will be handle on the gelato side while listening the event and building the right payload for `executeNextTx`. not need to track imo this `executeNextTx` in this piece of logic
 
maybe something to actually think about is if we would like to queue a tx if there is one pending or how to handle that. say, there is an existing transfer before which could bring you below the buffer etc

**Comment by petrovska-petro** [comment]
> looks like forked block height is again too far away? can we maybe fork on latest by default for now?

pinged always latest now 3dc7397182d7194b33e7722ca800bcaa61e7b01c

**Comment by gosuto-inzasheru** [comment]
pinned*

File: test/TopupEUReTest.t.sol (hunk 2 above)
Comment (petrovska-petro) [reply test/TopupEUReTest.t.sol]: included in 5393e3edd431477b0c28b52d06a8ba081d2837a7

**Pull request merged**